feat: add C implementation for `stats/base/dists/planck/entropy`

yuvimittal opened the pull request:
Resolves #4897.

## Description

> What is the purpose of this pull request?

This pull request: 

-   Add C implementation for @stdlib/stats/base/dists/planck/entropy
## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #4897 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

yuvimittal committed:
Add C implementation for @stdlib/stats/base/dists/planck/entropy

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

yuvimittal committed:
refactor random number generation

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/benchmark/benchmark.js:
```diff
@@ -21,7 +21,8 @@
 // MODULES //
 
 var bench = require( '@stdlib/bench' );
```
Changes in this file should be reverted.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/README.md:
```diff
@@ -120,6 +120,97 @@ for ( i = 0; i < lambda.length; i++ ) {
+
+#### stdlib_base_dists_planck_entropy( lambda )
+
+Evaluates the entropy for planck distribution.
```
```suggestion
Returns the [differential entropy][entropy] of a Planck distribution with shape parameter `lambda` (in [nats][nats]).
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/README.md:
```diff
@@ -120,6 +120,97 @@ for ( i = 0; i < lambda.length; i++ ) {
+#include <stdlib.h>
+#include <stdio.h>
+
+static double random_uniform( const double min, const double max ) {
+    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
+    return min + ( v*(max-min) );
+}
+int main( void ) {
+    double lambda;
+    double y;
+    int i;
+
+    for ( i = 0; i < 25; i++ ) {
+        lambda = random_uniform( 0.1, 5.0 );
+        y = stdlib_base_dists_planck_entropy( lambda );
+        printf( "λ: %lf, H(X;λ): %lf\n", lambda, y );
+    }
+}
```
```suggestion
static double random_uniform( const double min, const double max ) {
    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
    return min + ( v*(max-min) );
}

int main( void ) {
    double lambda;
    double y;
    int i;

    for ( i = 0; i < 25; i++ ) {
        lambda = random_uniform( 0.1, 5.0 );
        y = stdlib_base_dists_planck_entropy( lambda );
        printf( "λ: %lf, H(X;λ): %lf\n", lambda, y );
    }
}
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/include/stdlib/stats/base/dists/planck/entropy.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the entropy for planck distribution.
```
```suggestion
* Evaluates the differential entropy for a Planck distribution with shape parameter `lambda`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+*
+* @private
+* @param {number} lambda - shape parameter of the Planck distribution
+* @returns {number} evaluated differential entropy in nats
```
```suggestion
* @returns {number} differential entropy in nats
```

Planeshifter reviewed (changes requested):
Thanks for this PR, @yuvi-mittal! 

Would you mind looking into addressing the comments and CI lint failures, please?

yuvimittal committed:
feat: change years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

yuvimittal committed:
feat: change years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

yuvimittal committed:
feat: change years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

yuvimittal commented:
@Planeshifter , Thanks for reviewing , i have made the necessary changes !

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/README.md:
```diff
@@ -120,6 +120,98 @@ for ( i = 0; i < lambda.length; i++ ) {
+
+The function accepts the following arguments:
+
+-   **lambda**: `[in] double` shape parameter
```
```suggestion
-   **lambda**: `[in] double` shape parameter.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+// MAIN //
+
+/**
+* Evaluates the differential entropy for a Planck distribution with shape parameter `lambda`.
```
```suggestion
* Returns the differential entropy for a Planck distribution with shape parameter `lambda`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/include/stdlib/stats/base/dists/planck/entropy.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the differential entropy for a Planck distribution with shape parameter `lambda`.
```
```suggestion
* Returns the differential entropy for a Planck distribution with shape parameter `lambda`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/planck/entropy/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+* Evaluates the differential entropy for a Planck distribution with shape parameter `lambda`.
+*
+* @private
+* @param {number} lambda - shape parameter of the Planck distribution
```
```suggestion
* @param {PositiveNumber} lambda - shape parameter
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
chore: revert changes to `benchmark.js`

Planeshifter committed:
bench: refactor native benchmark to use uniform array

Planeshifter reviewed (approved):
Let's get this one in. Thanks, @yuvi-mittal!

I had to manually revert the changes to `benchmark.js` and also updated `benchmark.native.js` to use `random/base/uniform/array`. Please make these changes in your other open PRs for the Planck distribution as well.
